syntax = "proto3";

package torrent_p2p;

// Common message wrapper for all gossip types
message GossipMessage {
  // Metadata for all messages
  string message_id = 1;
  uint64 timestamp = 2;
  string source_ip = 3;
  uint32 source_port = 4;
  uint64 lamport_timestamp = 5;

  // Different message types
  oneof message_type {
    ReputationMessage reputation = 10;
    HeartbeatMessage heartbeat = 11;
    IndexSyncMessage index_sync = 12;
    // Add more message types here as needed
  }
}

// Reputation message
message ReputationMessage {
  string subject_ip = 1;
  uint32 subject_port = 2;
  int32 reputation = 3;
}

//Bootstrap Heartbeat Message
message HeartbeatMessage {
    enum Type {
        PING = 0;
        PONG = 1;
        INDEX_PING = 2;
        INDEX_PONG = 3;
    }
    Type type = 1;
    int64 timestamp = 2;
}

// Index Sync Message for synchronizing index data between nodes
message IndexSyncMessage {
    enum Type {
        REQUEST_FULL = 0;
        RESPONSE_FULL = 1;
        REQUEST_PARTIAL = 2;
        RESPONSE_PARTIAL = 3;
    }
    Type type = 1;
    int64 timestamp = 2;
    // For partial sync requests, specify the time range
    int64 start_timestamp = 3;
    int64 end_timestamp = 4;
}